### PR TITLE
Compilation fixes for DMD 2.083.0

### PR DESCRIPTION
**Pull request by dhasenan**
* `delete` is deprecated
* `!<>` is removed
* implicit string concatenation is removed
* varargs were changed (`__va_argsave` removed)

### CODE REVIEW

**Comment by dhasenan** [comment]
I know that's somewhat behind the times, but little by little, right?

File: tango/text/convert/DateTime.d [hunk 1]
```diff
@@ -971,10 +971,10 @@ private DateTimeLocale EngUS =
                                            "Thursday", "Friday", "Saturday"],
         monthNames                      : ["January", "February", "March", "April",
                                            "May", "June", "July", "August", "September",
-                                           "October" "November", "December"],
+                                           "October", "November", "December"],
```
Comment (SiegeLord) [inline tango/text/convert/DateTime.d:974]: Lol...

File: tango/text/convert/Layout.d [hunk 2]
```diff
@@ -123,7 +123,7 @@ class Layout(T)
                 {
                     va_list ap;
 
-                    va_start(ap, __va_argsave);
+                    va_start(ap, result);
```
Comment (SiegeLord) [inline tango/text/convert/Layout.d]: I think this should be formatStr?

File: tango/text/convert/Layout.d [hunk 3]
```diff
@@ -218,7 +218,7 @@ class Layout(T)
                 {
                     va_list ap;
 
-                    va_start(ap, __va_argsave);
+                    va_start(ap, sink);
```
Comment (SiegeLord) [inline tango/text/convert/Layout.d]: I think this should be formatStr?

**Review by SiegeLord — CHANGES_REQUESTED** [review]
Thanks, that's a good start. I appreciate the effort. I've bad news though, it doesn't compile for me on Linux x64 due to an ICE in IEEE.d. I can get around it, but then running the unittester segfaulted. I got around that too, and then there's a bunch more of compile errors here and there ('cause Tango does use templates here and there, so errors won't show up until you instantiate things).

What OS are you testing it on?

For posterity, here are the commands I used:

Building:

```
make -j8 -f build/Makefile static-lib DC=dmd ARCH=x86_64
```
Result:
```
---
ERROR: This is a compiler bug.
Please report it via https://issues.dlang.org/enter_bug.cgi
with, preferably, a reduced, reproducible example and the information below.
DustMite (https://github.com/CyberShadow/DustMite/wiki) can help with the reduction.
---
DMD v2.081.1
predefs   DigitalMars Posix linux ELFv1 LittleEndian D_Version2 all D_SIMD D_InlineAsm_X86_64 X86_64 CRuntime_Glibc D_LP64 D_PIC assert D_HardFloat
binary    dmd
version   v2.081.1
config    /etc/dmd.conf
DFLAGS    -I/usr/include/dmd/phobos -I/usr/include/dmd/druntime/import -L-L/usr/lib/x86_64-linux-gnu -L--export-dynamic -fPIC
---
Internal error: dmd/backend/cg87.c 810
build/Makefile:131: recipe for target 'objs/tango/math/IEEE.o' failed
```

Workaround, disable asm (and add debug symbols, as this will get hairy):
```
make -j8 -f build/Makefile static-lib DC=dmd ARCH=x86_64 DCFLAGS="-version=TangoNoAsm -g"
```

That worked. Then, compile the unittester:

```
dmd build/src/unittester.d libtango-dmd.a
```
Then run it:
```
./unittester -c dmd  -o="-unittest -L-ltango-dmd -L-L. -debug=UnitTest" $(find tango -name '*.d' | grep -v win32 | grep -v solaris | grep -v darwin) | tee out
```

That segfaults inside Tango's text formatting. I'll stop here for now.

**Comment by SiegeLord** [comment]
Another note so I don't forget, the unitester segfault happens in Layout.process because `_ti` argument is NULL.

**Comment by dhasenan** [comment]
Looks like dub build and dub test doesn't do the same build and test as the makefile. Should probably fix that at some point.

Anyway, I fixed the problems you mentioned. Current problems:
* segfault in Variant (const-correctness issues? or my hamfistedness in making changes)
* Layout produces garbage for AA keys

Continuing to look into it.

**Comment by dhasenan** [comment]
As for the ICE, that seems to have been fixed by 2.083.0. Updated the title and commit info to match.

The minimal reproduction case for that ICE, if we want to try to work around it:
```D
int isNaN(real)
{
    return 0;
}

ulong getNaNPayload(real x)
{
    isNaN(x);
    ulong m = *cast(ulong*)&x;
    return 0;
}
```

**Review by SiegeLord — APPROVED** [review]

**Comment by SiegeLord** [comment]
Alright, this looks good to me (somehow some unittests were hanging, but I think that's some race condition in the unittester somehow). Is this all set to merge?

**Comment by dhasenan** [comment]
Sorry, yes, this should be good to go.

**Comment by SiegeLord** [comment]
Alright, looks good to me! Thanks @dhasenan, this was amazing work. I appreciate it.

**Pull request merged**